x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/HttpRequestExecutorService.java

```diff
@@ -0,0 +1,177 @@
+ * {@link org.apache.http.client.methods.HttpUriRequest} to set a timeout for how long this executor will wait
+ * attempting to execute a task (aka waiting for the connection manager to lease a connection). See
+ * {@link org.apache.http.client.config.RequestConfig.Builder#setConnectionRequestTimeout} for more info.
+ */
+class HttpRequestExecutorService extends AbstractExecutorService {
+    private static final Logger logger = LogManager.getLogger(HttpRequestExecutorService.class);
+
+    private final ThreadContext contextHolder;
+    private final String serviceName;
+    private final BlockingQueue<HttpTask> queue;
+    private final AtomicBoolean running = new AtomicBoolean(true);
+    private final CountDownLatch terminationLatch = new CountDownLatch(1);
+    private final HttpClientContext httpContext;
+    private final HttpClient httpClient;
+
+    @SuppressForbidden(reason = "properly rethrowing errors, see EsExecutors.rethrowErrors")
+    HttpRequestExecutorService(ThreadContext contextHolder, String serviceName, HttpClient httpClient) {
+        this(contextHolder, serviceName, httpClient, null);
+    }
+
+    @SuppressForbidden(reason = "properly rethrowing errors, see EsExecutors.rethrowErrors")
+    HttpRequestExecutorService(ThreadContext contextHolder, String serviceName, HttpClient httpClient, @Nullable Integer capacity) {
+        this.contextHolder = Objects.requireNonNull(contextHolder);
+        this.serviceName = Objects.requireNonNull(serviceName);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.httpContext = HttpClientContext.create();
+
+        if (capacity == null) {
+            this.queue = new LinkedBlockingQueue<>();
+        } else {
+            this.queue = new LinkedBlockingQueue<>(capacity);
+        }
+    }
+
+    /**
+     * Begin servicing tasks.
+     */
+    public void start() {
+        try {
+            while (running.get()) {
+                HttpTask task = queue.take();
+                if (task.shouldShutdown() || running.get() == false) {
+                    running.set(false);
+                    logger.debug(() -> format("Http executor service [%s] exiting", serviceName));
+                } else {
+                    executeTask(task);
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } finally {
+            terminationLatch.countDown();
+        }
+    }
+
+    private void executeTask(HttpTask task) {
+        try {
+            contextHolder.preserveContext(task).run();
+        } catch (Exception e) {
+            logger.error(format("Http executor service [%s] failed to execute request [%s]", serviceName, task), e);
+        }
+        EsExecutors.rethrowErrors(ThreadContext.unwrap(task));
+    }
+
+    public int queueSize() {
+        return queue.size();
+    }
+
+    @Override
+    public void shutdown() {
```

Comment:
Should users call `shutdownNow()` instead of `shutdown()` to get the list of queued tasks? If so this method should throw an `UnsupportedOperationException`

Reply:
I think the only code that would call these `shutdown*()` methods would be when the node exits. I was hoping we could rely on retries to handle a shutdown when there are tasks in the queue. That'll be in a separate PR that I'm working on. So we probably don't even need `shutdownNow()` but I figured it was easy enough to add it 🤷‍♂️ . I can have it throw `UnsupportedOperationException` if you prefer though.

Reply:
Good point. No need to change this now let's see how all the parts come together